pyterrier_dr/bge_m3.py

```diff
@@ -0,0 +1,146 @@
+    def doc_encoder(self, verbose=None, batch_size=None):
+        return BGEM3DocEncoder(self, verbose=verbose, batch_size=batch_size)
+    
+    # Can do dense, sparse and colbert encodings
+    def query_multi_encoder(self, verbose=None, batch_size=None, return_dense=True, return_sparse=True, return_colbert_vecs=True):
+        return BGEM3QueryEncoder(self, verbose=verbose, batch_size=batch_size, return_dense=return_dense, return_sparse=return_sparse, return_colbert_vecs=return_colbert_vecs)
+    def doc_multi_encoder(self, verbose=None, batch_size=None, return_dense=True, return_sparse=True, return_colbert_vecs=True):
+        return BGEM3DocEncoder(self, verbose=verbose, batch_size=batch_size, return_dense=return_dense, return_sparse=return_sparse, return_colbert_vecs=return_colbert_vecs)
+
+class BGEM3QueryEncoder(pt.Transformer):
+    def __init__(self, bge_factory: BGEM3, verbose=None, batch_size=None, max_length=None, return_dense=True, return_sparse=False, return_colbert_vecs=False):
+        self.bge_factory = bge_factory
+        self.verbose = verbose if verbose is not None else bge_factory.verbose
+        self.batch_size = batch_size if batch_size is not None else bge_factory.batch_size
+        self.max_length = max_length if max_length is not None else bge_factory.max_length
+
+        self.dense = return_dense
+        self.sparse = return_sparse
+        self.multivecs = return_colbert_vecs
+    
+    def encode(self, texts):
+        return self.bge_factory.model.encode(list(texts), batch_size=self.batch_size, max_length=self.max_length,
+                             return_dense=self.dense, return_sparse=self.sparse, return_colbert_vecs=self.multivecs)
+
+    def transform(self, inp: pd.DataFrame) -> pd.DataFrame:
+        assert all(c in inp.columns for c in ['query'])
+        
+        # check if inp is empty
+        if len(inp) == 0:
+            if self.dense:
+                inp = inp.assign(query_vec=[])
+            if self.sparse:
+                inp = inp.assign(query_toks=[])
+            if self.multivecs:
+                inp = inp.assign(query_embs_toks=[])
+            return inp
+
+        it = inp['query'].values
+        it, inv = np.unique(it, return_inverse=True)
+        if self.verbose:
+            it = pt.tqdm(it, desc='Encoding Queries', unit='query')
+        bgem3_results = self.encode(it)
+
+        if self.dense:
+            query_vec = [bgem3_results['dense_vecs'][i] for i in inv]
+            inp = inp.assign(query_vec=query_vec)
+        if self.sparse:
+            # for sparse convert ids to the actual tokens
+            query_toks = self.bge_factory.model.convert_id_to_token(bgem3_results['lexical_weights'])
+            inp = inp.assign(query_toks=query_toks)
+        if self.multivecs:
+            query_embs_toks = [bgem3_results['colbert_vecs'][i] for i in inv]
+            inp = inp.assign(query_embs_toks=query_embs_toks)
```

Comment:
Are these the embeddings themselves of the token IDs associated with each embedding?

If they're the embeddings, the column should be query_embs to match the colbert stuff.